### PR TITLE
[WIP] Compare sodiumoxide and rust_sodium

### PR DESCRIPTION
**Pull request by kpp**
```
cp -R rust_sodium/src/ sodiumoxide/
cp -R rust_sodium/rust_sodium-sys/src/ sodiumoxide/libsodium-sys/
cp -R rust_sodium/rust_sodium-sys/build.rs sodiumoxide/libsodium-sys/build.rs
```

We should decide which features we would like to integrate:
1. for 100% sure
2. 50/50
3. never

I would like people to discuss features right in the code using comments. Goto `Files changed` and feel free to discuss the changes. We will integrate them part by part, once a feature is merged into master I will rebase this PR on master and the total diff will become smaller and the old comments will disappear.

`sodiumoxide` issue: https://github.com/sodiumoxide/sodiumoxide/issues/192
`rust_sodium` issue: https://github.com/maidsafe/rust_sodium/issues/67

### CODE REVIEW

File: src/crypto/aead/aead_macros.rs [hunk 1]
```diff
@@ -41,8 +41,8 @@ new_type! {
 /// `gen_key()` randomly generates a secret key
 ///
 /// THREAD SAFETY: `gen_key()` is thread-safe provided that you have
-/// called `sodiumoxide::init()` once before using any other function
-/// from sodiumoxide.
+/// called `rust_sodium::init()` once before using any other function
```
Comment (kpp) [inline src/crypto/aead/aead_macros.rs]: Never

File: src/crypto/aead/aead_macros.rs [hunk 2]
```diff
@@ -63,19 +63,20 @@ pub fn gen_nonce() -> Nonce {
...
-        $seal_name(
+        let _ = $seal_name(
             c.as_mut_ptr(),
             &mut clen,
             m.as_ptr(),
             m.len() as c_ulonglong,
             ad_p,
             ad_len,
-            0 as *mut _,
+            ptr::null_mut(),
```
Comment (kpp) [inline src/crypto/aead/aead_macros.rs]: 100%

File: src/crypto/aead/aead_macros.rs [hunk 3]
```diff
@@ -63,19 +63,20 @@ pub fn gen_nonce() -> Nonce {
...
 pub fn seal(m: &[u8], ad: Option<&[u8]>, n: &Nonce, k: &Key) -> Vec<u8> {
-    let (ad_p, ad_len) = ad.map(|ad| (ad.as_ptr(), ad.len() as c_ulonglong)).unwrap_or((0 as *const _, 0));
+    let (ad_p, ad_len) = ad.map(|ad| (ad.as_ptr(), ad.len() as c_ulonglong))
+        .unwrap_or((ptr::null(), 0));
     let mut c = Vec::with_capacity(m.len() + TAGBYTES);
     let mut clen = c.len() as c_ulonglong;
 
     unsafe {
-        $seal_name(
+        let _ = $seal_name(
```
Comment (kpp) [inline src/crypto/aead/aead_macros.rs]: @dirvine Why did you use `let _ = `?

File: src/crypto/aead/aead_macros.rs [hunk 4]
```diff
@@ -6,18 +6,18 @@ macro_rules! aead_module (($seal_name:ident,
                            $noncebytes:expr,
                            $tagbytes:expr) => (
 
-#[cfg(not(feature = "std"))] use prelude::*;
 use libc::c_ulonglong;
 use randombytes::randombytes_into;
+use std::ptr;
```
Comment (kpp) [inline src/crypto/aead/aead_macros.rs]: Never since we want `no_std` feature

File: src/crypto/auth/auth_macros.rs [hunk 5]
```diff
@@ -42,26 +42,26 @@ pub fn gen_key() -> Key {
...
 pub fn authenticate(m: &[u8],
-                    k: &Key) -> Tag {
+                    &Key(ref k): &Key) -> Tag {
     unsafe {
         let mut tag = [0; TAGBYTES];
-        $auth_name(tag.as_mut_ptr(),
-                   m.as_ptr(),
-                   m.len() as c_ulonglong,
-                   k.0.as_ptr());
+        let _todo_use_result = $auth_name(tag.as_mut_ptr(),
```
Comment (kpp) [inline src/crypto/auth/auth_macros.rs]: @dirvine why `_todo_use_result`?

**Comment by Dylan-DPC** [comment]
@kpp oops thought i was on the wrong issue. can you check why the appveyor build](https://ci.appveyor.com/project/Dylan-DPC/sodiumoxide/builds/20272968/job/rp00qofjoqk2he6s) is failing?

**Comment by kpp** [comment]
> can you check why the appveyor build is failing?

Because I didn't copy `Cargo.toml` files =) This PR is not supposed to be merged. It is a PR for discussion only.

**Comment by Dylan-DPC** [comment]
@kpp yeah i'm aware. i got an email saying this build is failing and I thought i was in the process of merging this. Realised it later. :smile:

File: src/crypto/aead/aead_macros.rs [hunk 6]
```diff
@@ -6,18 +6,18 @@ macro_rules! aead_module (($seal_name:ident,
...
 
 /// Number of bytes in a `Key`.
-pub const KEYBYTES: usize = $keybytes as usize;
```
Comment (kpp) [inline src/crypto/aead/aead_macros.rs]: 100%

File: src/crypto/aead/chacha20poly1305.rs [hunk 7]
```diff
@@ -1,41 +1,43 @@
...
-use ffi::{crypto_aead_chacha20poly1305_encrypt,
-          crypto_aead_chacha20poly1305_decrypt,
-          crypto_aead_chacha20poly1305_encrypt_detached,
-          crypto_aead_chacha20poly1305_decrypt_detached,
-          crypto_aead_chacha20poly1305_KEYBYTES,
-          crypto_aead_chacha20poly1305_NPUBBYTES,
-          crypto_aead_chacha20poly1305_ABYTES
+
+use ffi::{
+    crypto_aead_chacha20poly1305_ABYTES, crypto_aead_chacha20poly1305_KEYBYTES,
```
Comment (kpp) [inline src/crypto/aead/chacha20poly1305.rs]: 50/50

File: src/crypto/aead/chacha20poly1305.rs [hunk 8]
```diff
@@ -1,41 +1,43 @@
...
-             crypto_aead_chacha20poly1305_NPUBBYTES,
-             crypto_aead_chacha20poly1305_ABYTES);
+aead_module!(
+    crypto_aead_chacha20poly1305_encrypt,
+    crypto_aead_chacha20poly1305_decrypt,
+    crypto_aead_chacha20poly1305_encrypt_detached,
+    crypto_aead_chacha20poly1305_decrypt_detached,
+    crypto_aead_chacha20poly1305_KEYBYTES as usize,
+    crypto_aead_chacha20poly1305_NPUBBYTES as usize,
+    crypto_aead_chacha20poly1305_ABYTES as usize
```
Comment (kpp) [inline src/crypto/aead/chacha20poly1305.rs]: `as usize` 100%

File: src/crypto/aead/chacha20poly1305.rs [hunk 9]
```diff
@@ -1,41 +1,43 @@
...
 };
-aead_module!(crypto_aead_chacha20poly1305_encrypt,
-             crypto_aead_chacha20poly1305_decrypt,
-             crypto_aead_chacha20poly1305_encrypt_detached,
-             crypto_aead_chacha20poly1305_decrypt_detached,
-             crypto_aead_chacha20poly1305_KEYBYTES,
-             crypto_aead_chacha20poly1305_NPUBBYTES,
-             crypto_aead_chacha20poly1305_ABYTES);
+aead_module!(
+    crypto_aead_chacha20poly1305_encrypt,
```
Comment (kpp) [inline src/crypto/aead/chacha20poly1305.rs]: format 50/50

File: src/crypto/auth/mod.rs [hunk 10]
```diff
@@ -23,19 +23,19 @@
...
 //!
-//! ------------------------------------------------------------
-//! |crypto_auth              |primitive        |BYTES|KEYBYTES|
-//! |-------------------------|-----------------|-----|--------|
-//! |crypto_auth_hmacsha256   |HMAC_SHA-256     |32   |32      |
-//! |crypto_auth_hmacsha512256|HMAC_SHA-512-256 |32   |32      |
-//! |crypto_auth_hmacsha512   |HMAC_SHA-512     |64   |32      |
-//! ------------------------------------------------------------
+//! -----------------------------------------------------------------
+//! |`crypto_auth`               |primitive          |BYTES|KEYBYTES|
```
Comment (kpp) [inline src/crypto/auth/mod.rs]: 100%

**Comment by kpp** [comment]
Merged huges PRs: https://github.com/sodiumoxide/sodiumoxide/pull/285  https://github.com/sodiumoxide/sodiumoxide/pull/281

Before : +6,124 −2,104
After: +2,121 −1,776

File: src/crypto/auth/auth_macros.rs [hunk 11]
```diff
@@ -156,3 +160,122 @@ mod bench_m {
 }
 
 ));
+
```
Comment (kpp) [inline src/crypto/auth/auth_macros.rs]: Ignore it since it is a fix in sodiumoxide. See https://github.com/sodiumoxide/sodiumoxide/issues/272

**Comment by mitchtbaum** [comment]
anything new to report?

**Comment by kpp** [comment]
Seems like we drew almost all useful changes.

**Pull request closed without merging**